src/common/snippets/src/lowered/expression_factory.cpp

```diff
@@ -113,14 +117,20 @@ ExpressionPtr LinearIR::ExpressionFactory::create(const std::shared_ptr<op::Loop
     return expr;
 }
 
-ExpressionPtr LinearIR::ExpressionFactory::create(const std::shared_ptr<ov::Node>& n, const std::vector<PortConnectorPtr>& inputs) {
+ExpressionPtr LinearIR::ExpressionFactory::create(const std::shared_ptr<ov::Node>& n,
+                                                  const std::vector<PortConnectorPtr>& inputs,
+                                                  const LinearIR& linear_ir) {
     OPENVINO_ASSERT(!ov::is_type<ov::op::v0::Parameter>(n) &&
                     !ov::is_type<ov::op::v0::Result>(n),
                     "Expression builder with inputs doesn't support Result and Parameter");
-    auto expr = std::shared_ptr<Expression>(new Expression(n));
+    auto expr = std::shared_ptr<Expression>(new Expression(n, linear_ir.m_shape_infer_factory));
     init_expression_inputs(expr, inputs);
     create_expression_outputs(expr);
     expr->validate();
+    // todo: here we blindly synchronize input shapes from parent and child. Remove this when shapes will be stored in
+    //  port connector itself
+    if (linear_ir.m_shape_infer_factory)
```

Comment:
`m_shape_inference` must always be initialized, since we call `make_shape_inference` directly in the initializer list. If the factory failed to initialize `shapeInfer` object, then it throws an exception.
So the only way `m_shapeInference` could be nullptr is when one of the builder deliberately returned nullptr. In this case we'll throw an explicit exception from `updateShapes()` `catch` block.
So I think we should be fine here.